R/formulas.R

```diff
@@ -0,0 +1,173 @@
+#' Fit a list of formulas
+#'
+#' \code{fit_with()} is a pipe-friendly tool that applies a list of
+#' formulas to a fitting function such as \code{\link[stats]{lm}()}.
+#' The list of formulas is typically created with \code{\link{formulas}}().
+#'
+#' Assumes that \code{.f} takes the formula either as first argument
+#' or as second argument if the first argument is \code{data}.  Most
+#' fitting functions should fit these requirements.
+#' @param data A dataset used to fit the models.
+#' @param .f A fitting function such as \code{\link[stats]{lm}()},
+#'   \code{\link[lme4]{lmer}}() or \code{\link[rstanarm]{stan_glmer}()}.
+#' @param .formulas A list of formulas specifying a model.
+#' @param ... Additional arguments passed on to \code{.f}
+#' @seealso \code{\link{formulas}}()
+#' @export
+#' @examples
+#' # fit_with() is typically used with formulas().
+#' disp_fits <- mtcars %>% fit_with(lm, formulas(~disp,
+#'   additive = ~drat + cyl,
+#'   interaction = ~drat * cyl,
+#'   full = add_predictors(interaction, ~am, ~vs)
+#' ))
+#'
+#' # The list of fitted models is named after the names of the list of
+#' # formulas:
+#' disp_fits$full
+#'
+#' # Additional arguments are passed on to .f
+#' mtcars %>% fit_with(glm, list(am ~ disp), family = binomial)
+fit_with <- function(data, .f, .formulas, ...) {
+  args <- list(...)
+
+  # Shadowing data to avoid getting the whole raw SEXP structure in
+  # the calls captured by the fitting function (which gets displayed
+  # in print methods)
+  args$data <- quote(data)
+
+  map(.formulas, function(formula) {
+    do.call(".f", c(list(formula), args))
```

Comment:
yes but we'll need to adjust the environment used in `invoke()` then:

```r
invoke <- function(.f, .x = NULL, ...) {
  do.call(.f, c(.x, list(...)), envir = parent.frame())
}
```

How should we handle derived functions like `invoke_map()`? Probably no other way than adding `.env` as argument:

```r
invoke <- function(.f, .x = NULL, ..., .env = parent.frame()) {
  do.call(.f, c(.x, list(...)), envir = .env)
}
```

What do you think?